src/frequenz/core/sentinels.py

```diff
@@ -0,0 +1,185 @@
+# License: MIT
+# Copyright © 2021-2022 Tal Einat
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+# Based on:
+# https://github.com/taleinat/python-stdlib-sentinels/blob/9fdf9628d7bf010f0a66c72b717802c715c7d564/sentinels/sentinels.py
+
+"""Create unique sentinel objects.
+
+This module provides a class, [`Sentinel`][frequenz.core.sentinels], which can be used
+to create unique sentinel objects as specified by [`PEP
+661`](https://peps.python.org/pep-0661/).
+"""
+
+
+import sys as _sys
+from threading import Lock as _Lock
+from typing import Self, cast
+
+__all__ = ["Sentinel"]
+
+
+# Design and implementation decisions:
+#
+# The first implementations created a dedicated class for each instance.
+# However, once it was decided to use Sentinel for type signatures, there
+# was no longer a need for a dedicated class for each sentinel value on order
+# to enable strict type signatures.  Since class objects consume a relatively
+# large amount of memory, the implementation was changed to avoid this.
+#
+# With this change, the mechanism used for unpickling/copying objects needed
+# to be changed too, since we could no longer count on each dedicated class
+# simply returning its singleton instance as before.  __reduce__ can return
+# a string, upon which an attribute with that name is looked up in the module
+# and returned.  However, that would have meant that pickling/copying support
+# would depend on the "name" argument being exactly the name of the variable
+# used in the module, and simply wouldn't work for sentinels created in
+# functions/methods.  Instead, a registry for sentinels was added, where all
+# sentinel objects are stored keyed by their name + module name.  This is used
+# to look up existing sentinels both during normal object creation and during
+# copying/unpickling.
+
+
+class Sentinel:
+    """Create a unique sentinel object.
+
+    Sentinel objects are used to represent special values, such as "no value" or "not
+    computed yet". They are used in place of [`None`][] to avoid ambiguity, since `None`
+    can be a valid value in some cases.
+
+    For more details, please check [`PEP 661`](https://peps.python.org/pep-0661/).
+
+    Example:
+        ```python
+        from frequenz.core.sentinels import Sentinel
+        from typing import assert_type
+
+        MISSING = Sentinel('MISSING')
+
+        def func(value: int | MISSING) -> None:
+            if value is MISSING:
+                assert_type(value, MISSING)
+            else:
+                assert_type(value, int)
```

Comment:
python3 example.py
Traceback (most recent call last):
  File "/home/marenz/frequenz/frequenz-client-dispatch-python/example.py", line 8, in <module>
    def func(value: int | MISSING) -> None:
                    ~~~~^~~~~~~~~
TypeError: unsupported operand type(s) for |: 'type' and 'Sentinel'